### PR TITLE
Add custom syntax checking using ast-grep

### PR DESCRIPTION
**Pull request by skejserjensen**
This PR adds custom rules for syntax checking of the code using [`ast-grep`](https://ast-grep.github.io/) and makes GitHub Actions automatic use these rules to check all PRs and the main branch.

### CODE REVIEW

File: rules/leftover_debugging.yml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+id: leftover_debugging
+message: Leftover debug constructs or inforamtion.
```
Comment (CGodiksen) [inline rules/leftover_debugging.yml:2]: "information" instead.

File: rules/missing_punctuation.yml [hunk 2]
```diff
@@ -0,0 +1,14 @@
+id: missing_punctuation
```
Comment (CGodiksen) [inline rules/missing_punctuation.yml:1]: A rule could also be added to do this for log messages and error messages.

File: rules/test_not_named_test.yml [hunk 3]
```diff
@@ -0,0 +1,14 @@
+id: test_not_named_test
+message: Test function not named test_.
+severity: error
+language: Rust
+# Test functions (annotated with #[test] or #[tokio::test]) be named test_ etc.
```
Comment (CGodiksen) [inline rules/test_not_named_test.yml:5]: Maybe `must be named test_ etc.` instead.

File: rules/missing_punctuation.yml (hunk 2 above)
Comment (CGodiksen) [inline rules/missing_punctuation.yml:1]: We could also consider rules for when doc comments are necessary. Currently all modules, structs, struct fields, methods (except `new()`), and functions have doc comments. 

Note that this rule might not be ideal since we might have very small functions and test utility functions that we do not think need doc comments.

Comment (CGodiksen) [inline rules/missing_punctuation.yml:1]: I think it would be hard to enforce the grouped structure of our imports with a rule in general but maybe we could add a rule to make sure `use super::*;` is always first in the test imports and always followed by an empty line. Quickly looking through the code it seems we are a little inconsistent here.

**Review by CGodiksen — CHANGES_REQUESTED** [review]

File: rules/leftover_debugging.yml (hunk 1 above)
Comment (skejserjensen) [reply rules/leftover_debugging.yml:2]: Fixed.

File: rules/missing_punctuation.yml (hunk 2 above)
Comment (skejserjensen) [reply rules/missing_punctuation.yml:1]: Add to task.

File: rules/test_not_named_test.yml (hunk 3 above)
Comment (skejserjensen) [reply rules/test_not_named_test.yml:5]: Fixed.

File: rules/missing_punctuation.yml (hunk 2 above)
Comment (skejserjensen) [reply rules/missing_punctuation.yml:1]: Added to task.

Comment (skejserjensen) [reply rules/missing_punctuation.yml:1]: Added to task.

**Review by CGodiksen — APPROVED** [review]

**Review by chrthomsen — APPROVED** [review]
